sdk/zksend/src/wallet/channel/index.ts

```diff
@@ -12,93 +12,81 @@ export const DEFAULT_STASHED_ORIGIN = 'https://getstashed.com';
 
 export { StashedRequest, StashedResponse };
 
-interface StashedPopupOptions {
+interface StashedPopupOptions<T extends StashedRequestData['type']> {
 	origin?: string;
 	name: string;
+	type: T;
 }
 
-export class StashedPopup {
-	#id: string;
-	#origin: string;
+export class StashedPopup<T extends StashedRequestData['type']> {
+	#popup: Window | null;
 	#name: string;
+	#origin: string;
+	#id: string;
+	#interval: ReturnType<typeof setInterval> | null = null;
+	#type: string;
+	#resolve: (data: StashedResponseTypes[T]) => void;
+	#reject: (error: Error) => void;
+	#promise: Promise<StashedResponseTypes[T]>;
+
+	constructor({ origin = DEFAULT_STASHED_ORIGIN, name, type }: StashedPopupOptions<T>) {
+		const { promise, resolve, reject } = withResolvers();
+		this.#promise = promise;
+		this.#resolve = resolve;
+		this.#reject = reject;
 
-	#close?: () => void;
-
-	constructor({ origin = DEFAULT_STASHED_ORIGIN, name }: StashedPopupOptions) {
 		this.#id = crypto.randomUUID();
-		this.#origin = origin;
-		this.#name = name;
-	}
+		this.#popup = window.open('about:blank', '_blank');
 
-	async createRequest<T extends StashedRequestData>(
-		request: T,
-	): Promise<StashedResponseTypes[T['type']]> {
-		const popup = window.open('about:blank', '_blank');
-
-		if (!popup) {
+		if (!this.#popup) {
 			throw new Error('Failed to open new window');
 		}
 
-		const { promise, resolve, reject } = withResolvers<StashedResponseTypes[T['type']]>();
-
-		let interval: NodeJS.Timer | null = null;
-
-		function cleanup() {
-			if (interval) {
-				clearInterval(interval);
-			}
-			window.removeEventListener('message', listener);
-		}
-
-		const listener = (event: MessageEvent) => {
-			if (event.origin !== this.#origin) {
-				return;
-			}
-			const { success, output } = safeParse(StashedResponse, event.data);
-			if (!success || output.id !== this.#id) return;
-
-			cleanup();
-
-			if (output.payload.type === 'reject') {
-				reject(new Error('User rejected the request'));
-			} else if (output.payload.type === 'resolve') {
-				resolve(output.payload.data as StashedResponseTypes[T['type']]);
-			}
-		};
-
-		this.#close = () => {
-			cleanup();
-			popup?.close();
-		};
-
-		window.addEventListener('message', listener);
+		this.#origin = origin;
+		this.#name = name;
+		this.#type = type;
```

Comment:
You technically don't need to change this, you could keep all of this in `send` and just reference the popup